### ISSUE TITLE
Review Accessibility Object Model

### ISSUE THREAD

**Issue by travisleithead** [w3ctag/design-reviews#134]
See: https://a11y-api.github.io/a11y-api/spec/

**Comment by marcoscaceres** [comment]
I would hold a little bit on this. We are still working through quite a few issues and I feel it's still premature to do a TAG review.

**Comment by triblondon** [comment]
Picked up at F2F Tokyo, agreed to defer until next F2F unless you want to bump it.

**Comment by triblondon** [comment]
@marcoscaceres Are we still holding on this?

**Comment by marcoscaceres** [comment]
Yes. But expect a big announcement in the next few days.

**Comment by slightlyoff** [comment]
Suggesting we close and re-open when there's an update. Thoughts, @marcoscaceres?

**Comment by dbaron** [comment]
Also, is there a URL to the current spec?  Original URL is 404.

**Comment by LJWatson** [comment]
@dbaron the [Phase 1 spec](https://wicg.github.io/aom/spec/), and the [repo](https://github.com/WICG/aom)

**Comment by marcoscaceres** [comment]
Deferring to @alice, as Alice will know best when the spec will be ready for review.

**Comment by triblondon** [comment]
@marcoscaceres @alice are you ready for this review yet?

**Comment by alice** [comment]
Phase 1 is definitely ready for a look, and is available to experiment with in Chrome Beta per these instructions: https://groups.google.com/a/chromium.org/forum/#!topic/chromium-accessibility/UCfWokyvoj0

Your thoughts on the later phases (Phase 2 has a [WIP draft](https://wicg.github.io/aom/spec/phase2.html); later phases are sketched out in concept only in the [explainer](https://github.com/WICG/aom/blob/gh-pages/explainer.md)) would also be welcome.

**Comment by travisleithead** [comment]
Super. We'll take a look at Phase 2 in one of our upcoming teleconferences. Thanks for getting back to us!

**Comment by torgo** [comment]
Taken up at Nice f2f.

**Comment by torgo** [comment]
Discussed at [London F2F](https://github.com/w3ctag/meetings/tree/gh-pages/2018/01-london) - we are doing work here. Discussing between @travisleithead and @cynthia to set up a special call and ask @alice to join.

**Comment by torgo** [comment]
Will try to have a session slot at a time that makes sense for our f2f.

**Comment by torgo** [comment]
Picked up with @alice as guest dialing in for for our f2f.

**Comment by torgo** [comment]
Alice pointed us to the latest iteration of the explainer: https://github.com/WICG/aom/blob/HEAD/explainer.md

**Comment by travisleithead** [comment]
I'm very supportive of what I've seen from [the gist](https://gist.github.com/alice/15ac1f8f147bc0648957ccd52e6de928), specifically, not having special accessibility events, and instead surfacing requests that come through an AT as normal (indistinguishable, and thus not special-cased) keyboard events.

..and the [explainer](https://github.com/WICG/aom/blob/HEAD/explainer.md) is looking good as well; love the helpful code samples and usage descriptions! I just have a few comments at this point.

* In the section on [reflecting ARIA attributes](https://github.com/WICG/aom/blob/HEAD/explainer.md#reflecting-aria-attributes), I wonder if there will be any planned semantic difference between the exposed aria attributes in the DOM on HTML elements vs the ones proposed for ShadowRoot objects? I couldn't quite tell from the explainer, because the examples did not use any common aria attributes :-). E.g., would `ariaChecked` reflect as a string for HTML Elements, but as a Boolean for ShadowRoot objects? This really got interesting when reading the next section on [IDRef reflection](https://github.com/WICG/aom/blob/HEAD/explainer.md#use-case-2-setting-relationship-properties-without-needing-to-use-idrefs), because I wondered what `ariaActiveDescendant` would report on the ShadowRoot after being set to a value using `ariaActiveDescendantElement`. (I suspect certain aria reflecting attributes just wouldn't be present on the ShadowRoot, but it seems like having an `ariaActiveDescendantElement` assignment would be useful in other scenarios as well, so I'm hopeful these can be rationalized together to avoid inconsistencies between the ShadowRoot's set of exposed properties and the HTML Element's set.
* Another thing to consider for assigning default accessibility state to a custom element is [this trend toward providing default styles](https://github.com/w3c/webcomponents/issues/468), which make use of an options dictionary. This is likely only an option if there's no need to subsequently change the default values, which runs contrary to your example... so merely something to think about.
* The [precedence rules suggested here](https://github.com/WICG/aom/blob/HEAD/explainer.md#precedence-between-string-based-and-element-based-relationship-properties) seem overly complicated. Instead, I would rather not expose the primitive "string-reflecting" property (because these can always be obtained through `getAttribute` anyway), and just make the DOM property reflect in only one-direction (downstream from the content attribute). This is how event handler attributes work (e.g., `onclick`). Consider:
   ```js
      // 'aria-labelledby' content attribute reflects down to the IDL attribute... 
      // ... 'ariaLabelledBy', but changes to the IDL attribute are not reflected back.
      // Ex: assume <button aria-labelledby="id1">
      let elem = button.ariaLabelledBy; // Returns the element found by IDRef "id1"
      // (or null if nothing found).
      button.ariaLabelledBy = someOtherElement;
      // The content attribute is unchanged (getAttribute('aria-labelledby') is 
      // still 'id1', but the true value as *reflected* in the IDL attribute has changed.```

There may be a better way to do this, but having both IDL attributes with precedence rules seems undesirable to me.
* Based on the lastest Gist, I'm not sure any of the section on [accessible events](https://github.com/WICG/aom/blob/HEAD/explainer.md#user-action-events-from-assistive-technology) is still relevant. Maybe remove that section from the explainer?
* In the section on [virtual accessibility nodes](https://github.com/WICG/aom/blob/HEAD/explainer.md#virtual-accessibility-nodes), why are offset* properties included? These seem like a decidedly legacy [visual] styling concept that isn't terribly useful for an accessibility node?
* Also in the virtual accessibility nodes, won't virtual text nodes be needed? If so, perhaps the newer 'append', 'prepend', 'before', 'after' DOM APIs would be better for this non-legacy interface? (see [`ChildNode`](https://dom.spec.whatwg.org/#interface-childnode) and [`ParentNode`](https://dom.spec.whatwg.org/#interface-parentnode)
* The example of using the virtual accessibility tree hints at a cool feature: specialized accessibility nodes, e.g., like those for a role=table, which then have IDL for things like `colCount` and `rowcount`. Is there plan to make a hierarchical set of AccessibleNode objects? Are things like colCount and rowcount already provided via aria attributes?
* `ComputedAccessibleNode` still makes me uneasy. I wonder what unique set of state is missing that would be surfaced via such a node, but isn't already discoverable via planned "reflected" aria properties and their relationships in aria relational properties? I'm also not sure `getComputedStyle` is a great pattern to follow here...

**Comment by alice** [comment]
Thanks for the thorough review! Responding point by point, as there is a bit of a moving target problem here :smirk: 

> not having special accessibility events, and instead surfacing requests that come through an AT as normal (indistinguishable, and thus not special-cased) keyboard events.

After recent discussions, we also plan to surface InputEvents for the AT actions which would cause keyboard events to fire: https://gist.github.com/alice/570307ab24765de14837be1da49e1f98 ("Option 1" in that write-up.)

> ... assigning default accessibility state to a custom element ...

This is discussed in more detail in https://github.com/WICG/aom/issues/127. 

TL;DR
- We are moving away from having default state set on the Shadow Root, and towards having it be a Custom Element API.
- This should allow setting *both* static default (a la default styles) values shared by all instances of a particular element, *and* non-reflected dynamic states specific to a single instance.
- The shape of the Custom Element API is being discussed in https://github.com/w3c/webcomponents/issues/758 and I think will be discussed more at TPAC.

> ... would  `ariaChecked` reflect as a string for HTML Elements, but as a Boolean for ShadowRoot objects? 

Per above, we are likely dropping the ShadowRoot reflection piece. So, `ariaChecked` will reflect as a string for elements only.

> ... IDRef reflection ...
 
There has been a long discussion at https://github.com/whatwg/html/issues/3515 about IDREF reflection, with the result that I think we'll no longer have both string and element versions for relationship attributes.

> I'm not sure any of the section on accessible events is still relevant. Maybe remove that section from the explainer?

Agreed!

> In the section on virtual accessibility nodes, why are offset* properties included? These seem like a decidedly legacy [visual] styling concept that isn't terribly useful for an accessibility node?

Accessibility tree nodes have position attributes which are useful for multiple reasons:
- drawing outlines for sighted users to keep track of where AT focus is
- supporting "[touch exploration](https://support.google.com/accessibility/android/answer/6006598?hl=en)" modes

I'm not sure what you mean by "legacy" - what would the non-legacy version be?

> Also in the virtual accessibility nodes... perhaps the newer 'append', 'prepend', 'before', 'after' DOM APIs would be better for this non-legacy interface?

I'm not quite sure I follow the logic, sorry (but thanks for the pointers). Is this because append/prepend can take a DOMString? I'm not sure we've thought deeply about what virtual text nodes will be like.

Either way, happy to harmonise with newer APIs.

> Is there plan to make a hierarchical set of AccessibleNode objects? 

Good question. I'm not sure we've reached any kind of agreement on this yet - I think there were some competing options for how to support polymorphism in AccessibleNode objects.

> Are things like colCount and rowcount already provided via aria attributes?

Yup! Have a look at the supported properties on https://www.w3.org/TR/wai-aria-1.1/#table and https://www.w3.org/TR/wai-aria-1.1/#row

> I wonder what unique set of state is missing that would be surfaced via such a node, but isn't already discoverable via planned "reflected" aria properties and their relationships in aria relational properties?

For example:

```html
<button id="banana">Banana</button>
```

```js
console.log(banana.role);            // empty string - no ARIA role is set

let accessibleBanana = await getComputedAccessibleNode(banana);
console.log(accessibleBanana.role);  // "button"
```

Additionally, ARIA has a set of rules about when roles and attributes do and don't apply. 

For example, I can put an `aria-rowcount` attribute on a button element, but it won't be exposed in the accessibility tree, because `aria-rowcount` only applies to tables.

>  I'm also not sure `getComputedStyle` is a great pattern to follow here...

Would you be able to elaborate on this? Obviously we're not directly copying `getComputedStyle`, since it's an asynchronous API, but it was more to provide an analogy for the relationship between the specified semantic properties and the computed node - but definitely interested in pitfalls to avoid.

**Comment by travisleithead** [comment]
OK, coming back to this review! TPAC was a great chance to sync up, so thanks for your time there! Some point-by-point follow-ups on your last comments:

> After recent discussions, we also plan to surface InputEvents for the AT actions which would cause keyboard events to fire: https://gist.github.com/alice/570307ab24765de14837be1da49e1f98 ("Option 1" in that write-up.)

I participated in the discussion. Outcomes to surface as `input` events' [`inputType`](https://w3c.github.io/input-events/#events-inputevents) sounds great.

> The shape of the Custom Element API is being discussed in w3c/webcomponents#758 and I think will be discussed more at TPAC.

Yep, that was a spirited discussion. I'm hopeful that the `ElementInternals` will end-up with the right usage pattern, but that does sound like the right home for this stuff.

> Accessibility tree nodes have position attributes which are useful for multiple reasons:
> * drawing outlines for sighted users to keep track of where AT focus is
> * supporting "touch exploration" modes
>
> I'm not sure what you mean by "legacy" - what would the non-legacy version be?

I had this question answered when we chatted in-person. My understanding: if there will be a virtual a11y node, it must act as proxy for all the relevant state that would otherwise be on a regular Element. "Legacy" was just my way of grouping all pre-existing CSS-OM APIs. I hope we can help with the use-cases, but wondering if many of them can be addressed by the so-called "searchable invisible DOM" (which needs a better name BTW). Seems as if the virtual a11y node and related standarization of the a11y tree is still a way off, so not going to worry too much about it right now.

> ```js
> console.log(banana.role);            // empty string - no ARIA role is set
> ```

Ah, this is not what I expected. I thought the role IDL attribute might work more like input's `type` content attribute and IDL attribute (e.g., how the IDL attribute always returns the "computed" value). For custom-elements much of this might be made more sensical with the ElementInternals, but for other elements, I'd like to have the IDL attributes reflect/compute as much as possible--however I understand that might be much more complicated for certain types and there's an argument for consistency. But for role, I would really like to see that return a computed value!

**Comment by alice** [comment]
> I thought the role IDL attribute might work more like input's type content attribute and IDL attribute (e.g., how the IDL attribute always returns the "computed" value).

I'm not sure I follow. The `type` IDL attribute returns the value of the `type` content attribute, except where that value is missing or invalid, in which case it returns the default value.

Computing the role can be complex, for example the [ARIA 1.1 spec](https://www.w3.org/TR/wai-aria-1.1/) states:

> When an explicit or inherited role of `presentation` is applied to an element with the implicit semantic of a WAI-ARIA role that has required owned elements, in addition to the element with the explicit role of `presentation`, the user agent MUST apply an inherited role of `presentation` to any owned elements that do not have an explicit role defined. 

In terms of a code example, this means:

```html
<table role="presentation">
  <tr>
    <td>All</td>
    <td>these</td>
  </tr>
  <tr>
    <td>elements</td>
    <td>are</td>
  </tr>
  <tr>
    <td colspan="2">presentational</td>
  </tr>
</table>
```

Because a table must own rows, and rows must own cells, the `<table>` element having a role of `presentation` means that every `<tr>` and `<td>` also has a role of `presentation`, instead of `row` or `cell`. So computing the role can involve a tree walk.

**Comment by travisleithead** [comment]
At Paris F2F: Discussed this further (the computed accessibility node concept). There is a strong parallel to the problems we've seen in Style computation and read-back problems that have long been a sore spot for performance. In this respect, my idea to read the `role` value and have it imply a computation is probably not a good idea. We're assuming that the computation of a single a11y node in the tree involves potentially global (document-wide) searches to resolve things like element ids to element instances (for certain ARIA properties), etc., so this model of having an operation that returns a promise seems right. So, cudos on your approach. We do have some feedback on the approach to do the read-back, especially around locality, e.g., the requested a11y node should be more directly requested from the node of interest:

```js
banana.getA11yNode().then((n) => { console.log(n.role); });
```

**Comment by alice** [comment]
Thanks for the extra feedback!

I appreciate the suggestion about localising the `getComputedAccessibleNode()` call to the Node instance, but I was curious if you could provide some more context for that advice. We essentially intended to follow the model of `getComputedStyle()` more or less for consistency, so I'm interested in why we'd want to diverge, especially since AFAIK it'd be the first method on Node which returns a promise.

**Comment by domenic** [comment]
The getComputedStyle() design is very strange. My best guess for why it exists that way was that separate working groups produced the CSS APIs from the Node APIs, and they didn't want to collaborate and/or didn't have `partial interface` technology, so the CSS folks went with `getX(someY)` instead of `someY.getX()`.

I don't think there's anything special about promise return types that makes using them a cause for caution.

So I tend to agree that putting it as a method on `Node` makes more sense. It's a classic consistency vs. goodness tradeoff, in some sense. But in this case it's not a big pattern we're contemplating consistency consistent with---just a single weird old API from the early days of the web.

**Comment by annevk** [comment]
I suspect `getComputedStyle()` was there because there was this idea of [having different views](https://www.w3.org/TR/DOM-Level-2-Views/views.html) for the same document. So each view would have its own method.

**Comment by alice** [comment]
Thanks for the explanations! We'll definitely take that advice on board - keeping it promise based and moving it to Node.

**Comment by travisleithead** [comment]
Hey Alice,

Thanks so much for sticking with us through the [long] review! We appreciate all the time you've spent (and continue to sped) revising the various AOM docs and adjusting and reacting to everyone's feedback. We are looking forward to continuing the discussion.

Regarding this particular issue, we think we've reached the end of our general feedback--that the design is trending in a good direction, and that it might be more productive to work on a review for a specific aspect of AOM once the spec reaches a good milestone for review. So, what we'd like to ask of you is to file a new TAG design-review issue for each "level" of AOM that is ready for more specific review.

We look forward to continued involvement in the AOM project. Thanks so much and thanks for "flying TAG"!
